database/sqlite/sqlite_functions.c

```diff
@@ -2208,3 +2212,263 @@ void sql_load_node_id(RRDHOST *host)
+    if (unlikely(rc != SQLITE_OK))
+        goto bind_fail;
+    
+    param++;
+    rc = sqlite3_bind_int(res, 4, t_delta_first);
+    if (unlikely(rc != SQLITE_OK))
+        goto bind_fail;
+    
+    param++;
+    rc = sqlite3_bind_int(res, 5, t_delta_end);
+    if (unlikely(rc != SQLITE_OK))
+        goto bind_fail;
+    
+    param++;
+    rc = sqlite3_bind_text(res, 6, status, -1, SQLITE_STATIC);
+    if (unlikely(rc != SQLITE_OK))
+        goto bind_fail;
+
+    rc = execute_insert(res);
+    if (unlikely(rc != SQLITE_DONE))
+        error_report("Failed to store gap, rc = %d", rc);
+
+    rc = sqlite3_reset(res);
+    if (unlikely(rc != SQLITE_OK))
+        error_report("Failed to reset statement in gap store function, rc = %d", rc);
+
+    return rc;
+
+bind_fail:
+    error_report("Failed to bind parameter %d to store gap, rc = %d", param, rc);
+    rc = sqlite3_reset(res);
+    if (unlikely(rc != SQLITE_OK))
+        error_report("Failed to reset statement in gap store function, rc = %d", rc);
+    return rc;
+}
+
+/*
+ * Load a gap from the SQLite DB
+ */
+int sql_load_host_gap(RRDHOST *host)
+{
+    sqlite3_stmt *res = NULL;
+    int rc;
+
+    if (unlikely(!db_meta)) {
+        if (default_rrd_memory_mode == RRD_MEMORY_MODE_DBENGINE)
+            error_report("Database has not been initialized");
+        return SQLITE_ERROR;
+    }
+
+    rc = sqlite3_prepare_v2(db_meta, SQL_GET_HOST_GAPS, -1, &res, 0);
+    if (unlikely(rc != SQLITE_OK)) {
+        error_report("Failed to prepare statement to fetch gaps");
+        return SQLITE_ERROR;
+    };
+
+    rc = sqlite3_bind_text(res, 1, host->machine_guid, -1, SQLITE_STATIC);
+    if (unlikely(rc != SQLITE_OK)) {
+        error_report("Failed to bind host_id parameter to load gap information");
+        goto failed;
+    }
+
+    do {
+        rc = sqlite3_step(res);
+        switch (rc) {
+            case SQLITE_ROW:
+                if (likely(sqlite3_column_bytes(res, 0) == sizeof(uuid_t))) {
+                    set_host_gap(host, res);
+                    debug(D_REPLICATION, "%s: Setting %s host latest gap completed!", REPLICATION_MSG, host->hostname);
+                }
+                break;
+            case SQLITE_DONE:
+                set_host_gap(host, NULL);
+                debug(D_REPLICATION, "%s: Setting %s host gaps from SQLite completed with NO ROWs!", REPLICATION_MSG, host->hostname);
+                break;
+            default:
+                debug(D_REPLICATION, "%s: Setting %s host gaps from SQLite returned unexpected error code!", REPLICATION_MSG, host->hostname);
+                break;
+        }
+    } while (rc == SQLITE_ROW);
+
+failed:
+    if (unlikely(sqlite3_finalize(res) != SQLITE_OK))
+        error_report("Failed to finalize the prepared statement when loading all gaps information");
+    return rc;
+};
+
+/*
+ * Delete all gaps from the database
+ */
+int sql_delete_all_gaps(void)
+{
+    sqlite3_stmt *res = NULL;
```

Comment:
Two things, firstly @odynik , please, take a look in this [comment](https://github.com/netdata/netdata/pull/12556/files#r879026584).

Now @stelfrag , when we delete we are losing historical, should we not use another algorithm to keep this historical?